modules/renter/siafile/siafile.go

```diff
@@ -0,0 +1,180 @@
+package siafile
+
+import (
+	"encoding/base32"
+	"encoding/binary"
+	"fmt"
+	"reflect"
+	"sync"
+
+	"github.com/NebulousLabs/Sia/modules"
+	"github.com/NebulousLabs/Sia/types"
+	"github.com/NebulousLabs/fastrand"
+
+	"github.com/NebulousLabs/Sia/crypto"
+)
+
+type (
+	// SiaFile is the disk format for files uploaded to the Sia network.  It
+	// contains all the necessary information to recover a file from its hosts and
+	// allows for easy constant-time updates of the file without having to read or
+	// write the whole file.
+	SiaFile struct {
+		// metadata is the mostly static metadata of a SiaFile. The reserved
+		// size of the metadata on disk should always be a multiple of 4kib.
+		// The metadata is also the only part of the file that is JSON encoded
+		// and can therefore be easily extended.
+		metadata Metadata
+
+		// pubKeyTable stores the public keys of the hosts this file's pieces are uploaded to.
+		// Since multiple pieces from different chunks might be uploaded to the same host, this
+		// allows us to deduplicate the rather large public keys.
+		pubKeyTable []types.SiaPublicKey
+
+		// chunks are the chunks the file was split into.
+		chunks []Chunk
+
+		// utility fields. These are not persisted.
+		deleted     bool
+		erasureCode modules.ErasureCoder
+		mu          sync.RWMutex
+		uid         string
+	}
+
+	// Chunk represents a single chunk of a file on disk
+	Chunk struct {
+		// erasure code settings.
+		//
+		// erasureCodeType specifies the algorithm used for erasure coding
+		// chunks. Available types are:
+		//   0 - Invalid / Missing Code
+		//   1 - Reed Solomon Code
+		//
+		// erasureCodeParams specifies possible parameters for a certain
+		// erasureCodeType. Currently params will be parsed as follows:
+		//   Reed Solomon Code - 4 bytes dataPieces / 4 bytes parityPieces
+		//
+		erasureCodeType   [4]byte
+		erasureCodeParams [8]byte
+
+		// extensionInfo is some reserved space for each chunk that allows us
+		// to indicate if a chunk is special.
+		extensionInfo [16]byte
+
+		// pieces are the pieces of the file the chunk consists of.
+		pieces [][]Piece
+	}
+
+	// Piece represents a single piece of a chunk on disk
+	Piece struct {
+		KeyNonce   [4]byte            // nonce used for encrypting the piece
+		HostPubKey types.SiaPublicKey // public key of the host
+		MerkleRoot crypto.Hash        // merkle root of the piece
+	}
+)
+
+// New create a new SiaFile.
+func New(siaPath string, erasureCode modules.ErasureCoder, pieceSize, fileSize uint64) *SiaFile {
+	file := &SiaFile{
+		metadata: Metadata{
+			fileSize:  int64(fileSize),
+			masterKey: crypto.GenerateTwofishKey(),
+			pieceSize: pieceSize,
+			siaPath:   siaPath,
+		},
+		erasureCode: erasureCode,
+		uid:         base32.StdEncoding.EncodeToString(fastrand.Bytes(20))[:20],
```

Comment:
What is this used for? Why is it base 32? Why do you slice it to the first 20 bytes?

Reply:
I don't know. It's just the way we have done it before :P

Reply:
since the uid is not persisted ever, we can just use fastrand.Bytes, there's no need to encode it.

I've been wondering whether we should persist the uid, but I think that could cause problems.

And, if we aren't going to persist the uid but instead keep everything in memory, another approach would be to use a uid counter and just count up each time that we add a file, because all we are really looking for is uniqueness, and it doesn't need any sort of collision resistance or security from an external party.

Reply:
the downside to a counter of course is that you need to keep a counter somewhere.